Add LZ78 Compression and Decompression Implementation with Unit Tests

vaibhavsiingh opened the pull request:
This pull request introduces an implementation of the LZ78 compression algorithm along with its corresponding decompression functionality. Additionally, it includes unit tests.

### Implementation of the LZ78 algorithm:
* [`slp-eval-py/src/slp_eval/lz78.py`](diffhunk://#diff-0d064ff93c8ec1bf9b47156cd6325e992db07f09e219d056eeb1db1e704f3788R1-R48): Added a new `LZ78` class with methods `compress` and `decompress` to perform LZ78 compression and decompression. The implementation uses dictionaries to manage mappings and supports generic types for input data.

### Unit tests for LZ78:
* [`slp-eval-py/src/slp_eval/tests/lz78_test.py`](diffhunk://#diff-5a15b26788458a420b8f3baedf75a40c7c245b0ddcfe86e797938305bce15db5R1-R37): Added parameterized unit tests using `pytest` to verify the correctness of the LZ78 implementation. The tests cover edge cases like empty strings, single characters, repeated patterns, and complex repetitions.

vaibhavsiingh committed:
new: Add LZ78 compression and decompression.

vaibhavsiingh committed:
fix: handle final phrase correctly in LZ78 compress

vaibhavsiingh committed:
test: add comprehensive tests for LZ78 compression and decompression

AliaumeL commented on slp-eval-py/src/slp_eval/lz78.py:
```diff
@@ -0,0 +1,48 @@
+
+@dataclass
+class LZ78(Generic[A]):
+    """LZ78 Compressor and Decompressor."""
```
Can you put a reference in a comment to explain from where you got the implementation?

vaibhavsiingh replied:
I've pushed the requested changes:
Let me know if anything else needs adjusting!

AliaumeL reviewed (commented):
I believe that it would be better to have a class that *represents* a Lempel-Ziv compressd string rather than a class that contains two methods taking `self` as an argument and never using it. Note that if you are not using self, maybe you want to use decorators such as `@classmethod` or `@staticmethod`.

I would suggest that you write a class `LZ78String` that represents a compressed string.

```python
@dataclass
class LZ78String(Generic[A]):
  """ A compressed string using Lempel-Ziv """
  content : list[Tuple[int, A]]

  @staticmethod
  def from_list(s : list[A]) -> 'LZ78String[A]':
      pass

  def to_list(self) -> list[A]:
      pass
```

This has two advantages. First, you have better types because you can actually talk about compressed strings and not `list[tuple[A,int]]`. Second, the way to interact with your code is clear and discoverable: you export one type, and autocompletion will automatically guide users to `from_list` and `to_list` functions.

@vaibhavsiingh

vaibhavsiingh committed:
Refactor: LZ78String dataclass

- Represent compressed data as a dataclass field `content: List[Tuple[int, A]]` instead of separate methods
- Introduce static methods `_compress` and `_decompress` for core LZ78 logic
- Provide `from_list` and `to_list` for clear raw-to-compressed and back conversion
- Provide `from_codes` and `to_codes` for round-trip of stored codes, with input validation
- Add `to_slp` method to convert compressed content into an SLP representation
- Added tests for all the above methods

vaibhavsiingh committed:
fix: resolve linter error in compression_model.py and added source of lz78 compression algorithm in lz78.py

vaibhavsiingh committed:
Fix: Pyrefly linter error
